src/profile-logic/js-tracer.js

```diff
@@ -0,0 +1,340 @@
+    let timingRow = jsTracerTimingMap.get(rowName);
+    if (timingRow === undefined) {
+      timingRow = {
+        start: [],
+        end: [],
+        index: [],
+        label: [],
+        name: rowName,
+        length: 0,
+      };
+      jsTracerTimingMap.set(rowName, timingRow);
+    }
+    // Convert the timing to milliseconds.
+    const division = 1000;
+    if (end < start) {
+      throw new Error('end is less than the start');
+    }
+    timingRow.start.push(start / division);
+    timingRow.end.push(end / division);
+    timingRow.label.push(displayName);
+    timingRow.index.push(tracerEventIndex);
+    timingRow.length++;
+
+    if (timingRow.length > 1) {
+      const currStart = timingRow.start[timingRow.length - 1];
+      const currEnd = timingRow.end[timingRow.length - 1];
+      const prevEnd = timingRow.end[timingRow.length - 2];
+      if (currEnd < currStart) {
+        throw new Error(
+          `currEnd < currStart "${displayName} - ${currEnd} < ${currStart}"`
+        );
+      }
+      if (currStart < prevEnd) {
+        throw new Error(
+          `currStart < prevEnd "${displayName} - ${currStart} < ${prevEnd}"`
+        );
+      }
+      if (currEnd < prevEnd) {
+        throw new Error(
+          `currEnd < prevEnd "${displayName} - ${currEnd} < ${prevEnd}"`
+        );
+      }
+    }
+  }
+
+  // Determine the self time of the various events. These values are all the stack
+  // of "prefix" events. The event details will be pushed on to these stacks until
+  // they are clearly self time.
+  const prefixesStarts = [];
+  const prefixesEnds = [];
+  const prefixesEventIndexes = [];
+  let prefixesTip = -1;
+
+  // Go through all of the events. Each `if` branch is documented with a small diagram
+  // that includes a little bit of ascii art to help explain the step.
+  //
+  // Legend:
+  // xxxxxxxx - Already reported information, not part of the prefix stack.
+  // [======] - Some event on the stack that has not been reported.
+  // [prefix] - The prefix event to consider, the top of the prefix stack. This
+  //            could also be only part of an event, that has been split into multiple
+  //            pieces.
+  // [current] - The current event to add.
+
+  for (
+    let currentEventIndex = 0;
+    currentEventIndex < tracerEvents.length;
+    currentEventIndex++
+  ) {
+    const currentStart = tracerEvents.timestamps[currentEventIndex];
+    const durationRaw = tracerEvents.durations[currentEventIndex];
+    const duration = durationRaw === -1 ? 0 : durationRaw;
+    const currentEnd = currentStart + duration;
+
+    if (prefixesTip === -1) {
+      // Nothing has been added yet, add this "current" event to to the stack of prefixes.
+      prefixesTip = 0;
+      prefixesStarts[prefixesTip] = currentStart;
+      prefixesEnds[prefixesTip] = currentEnd;
+      prefixesEventIndexes[prefixesTip] = currentEventIndex;
+      continue;
+    }
+
+    while (prefixesTip >= 0) {
```

Comment:
I think we can replace this condition with `while (true)`, the current condition doesn't seem exercized at all. The only place where `prefixesTip` can be negative is when it's decremented below, and we're already breaking out of the loop in that case.

With `while(true)` it gives a clear hint that we'll break out of the loop with `break` statements.

Reply:
Alternatively we can move the block `if (prefixesTip === -1) {` to after the while loop.